devtools/startup/DevToolsStartup.jsm

```diff
@@ -1878,12 +1882,23 @@ function onRecordingStarted(recording) {
   recording.on("unusable", function(name, data) {
     const browser = getBrowser();
 
-    reloadAndClearRecordingState(browser, `about:replay?error=${data.why}`);
+    handleFailedRecording(browser, data.why);
   });
-  recording.on("finished", function(name, data) {
-    const browser = getBrowser();
+
+  recording.on("finished", function () {
+    // caching a reference to the browser before the process ends
+    browser = getBrowser();
+  });
+
+  recording.on("saved", function(name, data) {
     const recordingId = data.id;
 
+    if (!browser) {
+      // Be sure we received the finished event before the saved event
+      handleFailedRecording(browser, 'Recording processed out of order');
+      return;
+    }
+
     // When the submitTestRecordings pref is set we don't load the viewer,
```

Comment:
I realize we have a bit of a race condition now because a lot of this code should still ideally run as part of `finished`. Hmm

Reply:
Can you move everything from this line down to `reloadAndClearRecordingState(browser);` into `finished`? We'll also need to make sure this function still returns early if the `submitTestRecordings` pref is set.